### PR TITLE
Change Library to monorepo with better types 

### PR DESCRIPTION
**Pull request by louiss0**
There are huge changes to the system. I tried to do my best to respect the code that you have written while making sure that I added code that would lead to a better DX. But I decided to change everything. 
- Most of the code is written in just typescript. 
- I removed the Astro prefix from all components. 
- This repo is now a pnpm monorepo
- All components are typed differently
- There is a `types.ts` folder that controls all the types used for the project
- I removed each type for the individual components then replaced them with a single file called the `index.d.ts`

### CODE REVIEW

**Comment by jonathantneal** [comment]
Hey, thanks! I’m taking a look. I can appreciate the fresh approach.

Nit: In the refresh, you’ve changed the license without notice, so please do restore the license to CC0-1.0 before this gets merged.

File: packages/flow/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "editor.formatOnSave": true,
+  "editor.codeActionsOnSave": {
+    "source.fixAll": true
+  },
+  "cSpell.words": ["Astro"],
+  "prettier.documentSelectors": ["**/*.astro"]
+}
```
Comment (jonathantneal) [inline packages/flow/.vscode/settings.json:8]: This seems highly opinionated. Is this a recommended pattern for all Astro integrations or components?

File: packages/flow/.prettierrc.cjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      },
+    },
+    {
+      files: ['.*', '*.json', '*.md', '*.toml', '*.yml'],
+      options: {
+        useTabs: false,
+      },
+    },
+  ],
+};
```
Comment (jonathantneal) [inline packages/flow/.prettierrc.cjs]: Is it possible to use ESM here?

File: demo/src/pages/when.astro [hunk 3]
```diff
@@ -1,34 +1,33 @@
 ---
-import Document from '../layouts/Document.astro'
-import { When } from '@astropub/flow'
+import Document from "../layouts/Document.astro";
+import { When } from "@astropub/flow";
```
Comment (jonathantneal) [inline demo/src/pages/when.astro:3]: General Nit / Tip: For PR review, it can be helpful to follow the formatting conventions of the existing project. For example, when I’m PR’ing a project that uses 2-spaces or 4-spaces, then I use 2-spaces or 4-spaces, and if the project uses tabs then I use tabs. In this case, the changes to strings distract from the improvements you’re looking to highlight.

I am super interested to see the parts where you really wanted to change how things work.

File: demo/src/pages/index.astro [hunk 4]
```diff
@@ -1,22 +1,23 @@
...
+		<h2>Demonstrations</h2>
 		<ul>
 			<li>
 				<a href="./for/">The <code>&lt;For&gt;</code> component</a>
 			</li>
 			<li>
-				<a href="./switch/">The <code>&lt;Switch&gt;</code> component.</a>
+				<a href="./switch/"
+					>The <code>&lt;Switch&gt;</code> component.</a
+				>
```
Comment (jonathantneal) [inline demo/src/pages/index.astro:20]: This doesn't seem related to any of the changes you want to make. Out of curiosity, has a code formatter done this, or is this change something you believe improves the readability of the document?

File: demo/src/pages/when.astro [hunk 5]
```diff
@@ -1,34 +1,33 @@
...
-	<pre><code
->&lt;When &#x7B;true&#x7D;&gt;
+	<pre><code>&lt;When &#x7B;true&#x7D;&gt;
 	&lt;h2&gt;True&lt;/h2&gt;
 	&lt;Fragment slot=&quot;else&quot;&gt;
 		&lt;h2&gt;False&lt;/h2&gt;
 	&lt;/Fragment&gt;
 &lt;/When&gt;</code></pre>
-	<When {true}>
+	<When condition={true}>
```
Comment (jonathantneal) [inline demo/src/pages/when.astro:17]: Looking at the updated `when.astro`, I’m curious if this change is necessary. Should `<When>` require a specific key, or would it still evaluate all possible keys?

File: packages/flow/lib/types.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+type ScalarValues = string | number | bigint | boolean | symbol | null;
+
+interface HasForEachMethod {
+	forEach(callbackfn: (...args: Array<unknown>) => void, thisArg?: unknown): void;
+}
```
Comment (jonathantneal) [inline packages/flow/lib/types.ts:5]: This is really neat idea. What made you want to switch away from the built-in `Iterable` to this? Was it to support a looser range of values?

File: packages/flow/.vscode/settings.json (hunk 1 above)
Comment (louiss0) [reply packages/flow/.vscode/settings.json:8]: I used the official typescript community template for this project. I did not change anything about it. I'm willing to remove that stuff

File: packages/flow/.prettierrc.cjs (hunk 2 above)
Comment (louiss0) [reply packages/flow/.prettierrc.cjs]: I'm not sure

File: demo/src/pages/when.astro (hunk 3 above)
Comment (louiss0) [reply demo/src/pages/when.astro:3]: The changes were due to prettier. I think it's better for me to take it out of the project

File: demo/src/pages/index.astro (hunk 4 above)
Comment (louiss0) [reply demo/src/pages/index.astro:20]: I think it's due to the formatter. I'm going to change this

File: demo/src/pages/when.astro (hunk 5 above)
Comment (louiss0) [reply demo/src/pages/when.astro:17]: It does not require a specific key but the key has to be a string. The typing is 
```ts
Record<string, unknown>
```

I think forcing the developer to be explicit is a good thing

File: packages/flow/lib/types.ts (hunk 6 above)
Comment (louiss0) [reply packages/flow/lib/types.ts:5]: The problem is that you have to create an overload or a union type. This solution forces the developer to use a data structure. The best solution to this problem would have been to find the typing for the `for of` loop. But i couldn't find it. 
This solution makes it so that the type for data structure is inferred.

**Comment by louiss0** [comment]
Changed the branch. I did not tell you what happened in depth last time but I need you to know that this repo uses the test version of the library. I decided to leave the rest up to you. 

**This repo is a repo that uses pnpm's mono repo system.** 

- The `.pnpm-workspace.yaml` file is the file where you list all of your workspaces.
- The `@astropub` is the name of the monorepo
- When you use `workspace:{version of package}` as the name of the package you want to install it will install the package that is in the repo instead of the one that is on npm [pnpm workspace page](https://pnpm.io/workspaces)   

**I decided to get rid of the linting dependencies and change the MIT license.**

File: package.json [hunk 7]
```diff
@@ -1,32 +1,16 @@
 {
-	"name": "astro-flow-root",
-	"type": "module",
-	"version": "0.3.6",
-	"workspaces": [
-		"demo",
-		"packages/*"
-	],
```
Comment (jonathantneal) [inline package.json:8]: This changes the project from being a monorepo to _not_ a monorepo, which makes it a lot harder to test the demo.

File: demo/src/pages/for.astro [hunk 8]
```diff
@@ -10,95 +11,56 @@ import { For, Iterate, iterate } from '@astropub/flow'
...
-	</h2>
-	<pre><code
->&lt;ul&gt;
+	<pre><code>&lt;ul&gt;
 	&lt;For of=&lcub;[
 		&lcub; title: 'This is the first item.' &rcub;,
 		&lcub; title: 'This is the second item.' &rcub;,
 		&lcub; title: 'This is the last item.' &rcub;,
-	]&rcub;&gt;&lcub;function * (item) &lcub;
-		yield &lt;li&gt;&lcub;item.title&rcub; (component)&lt;/li&gt;
```
Comment (jonathantneal) [inline demo/src/pages/for.astro:81]: Why are all of the generators removed?

File: package.json (hunk 7 above)
Comment (louiss0) [reply package.json:8]: The code that is in there is moved to `pnpm-workspace.yaml`
 ```json
"workspaces": [
		"demo",
		"packages/*"
	],
```

Look in `pnpm-workspace.yaml` for this.
```yaml
packages:
  - "packages/**"
  - "demo/**"

```

File: demo/src/pages/for.astro (hunk 8 above)
Comment (louiss0) [reply demo/src/pages/for.astro:81]: You can't really use generators anymore. I mean you can put them in there but yield won't work properly. A generator is being used behind the scenes anyway. I'm not sure you looked at how iterate works but. It uses a generator.

File: demo/src/pages/when.astro (hunk 5 above)
Comment (jonathantneal) [reply demo/src/pages/when.astro:17]: First, I agree with the stronger typing. Know that the shorthand `{true}` becomes `true={true}`, meaning it still works even with the `Record<string, unknown>` typing.

At the time of this writing, there is a bug in the VSCode extension that prevents Astro from recognizing any shorthand properties. I would rather not cater to that bug, since it is not actually a bug in the product itself.

File: demo/src/pages/for.astro (hunk 8 above)
Comment (jonathantneal) [reply demo/src/pages/for.astro:81]: When I test the current version, the generators are working.
When I test this version, the generators are problematic. At one point they print `[object Object]`.

Comment (jonathantneal) [reply demo/src/pages/for.astro:81]: > yield won't work properly

That could be because you tried writing the iterate component in an Astro file. Astro has changed how components work a few times. Astro used to render components like a stream, skipping slots unless they were set to be rendered. Recently, they started rendering components really poorly from the inside out like React, including rendering slots regardless if they are applied. I am told this is acknowledged in a way that it will be fixed, which is why the current version supports both scenarios.

Comment (louiss0) [reply demo/src/pages/for.astro:81]: I changed `iterate()` so that it can work with just generators.

**Comment by louiss0** [comment]
Are there any problems with the code I have written so far?

**Comment by louiss0** [comment]
- The iterate function is fixed
- I changed the license from MIT to CC0-1.0

**Are there any other problems.**

I have not heard from you in a long time and you created another pull request? What is going on?

**Comment by jonathantneal** [comment]
Hi @louiss0, I have tagged you in the other PR, which I would like you to review.

> Are there any problems with the code I have written so far?

Well, yes. For example, your PR is documented as “Change Library to monorepo with better types” but it actually changes the library to not be a monorepo. The focus on typing is very unclear, and you made many changes totally unrelated to this. In your description, you talk about respecting the code, but then you add “_But I decided to change everything._” That makes it hard to see the gist of what you’re trying to accomplish.

In the PR that I tagged you in, it credits and attempts to apply the changes you suggested, without changing _everything_.

I very much want to partner with you here. It’s one of my favorite parts of open source. Again, it means a lot to me that you looked into the code and wanted to contribute. I wish I knew how to express that better.

I wanted to see if this PR could eventually take a shape that I could accept. It is still not there, which is why the alternative PR is up. The alternative tries to give you what you wanted without changing everything. There are some general issues stopping me from merging this PR as is. Most of them are side-effects from ‘changing everything’, and I will try to make them more explicit.

One nit: This PR makes it so users can’t contribute from a standard Node/NPM setup. I really like the accessibility of keeping things simple. That’s also one of the reasons I sometimes prefer JavaScript files over TypeScript files.

This may not seem relevant anymore, but at the time my alternative PR was made, your iterators were still not accepting generators. As my demos use iterators, I saw that you had changed them, but I had not seen a note from you explaining why you made this change. When I asked about it, you wrote that I can’t really use generators anymore, and I respectfully disagree with that. If this PR works with generators again, that deserves me giving it another look.

While I’ve listed out some issues with your PR, it’s only to make sure you know exactly why and how I’m hoping to get us the improvements you’ve brought up. As I wrote in my original reply to #1, I’m grateful you took the time to identify issues, and I want to get the typing exactly how you think it should be, as well as give you some credit. I’m also very sorry for any delays in my communication. I hope this feedback is helpful and not discouraging; either so that you can address the issues I’ve mentioned in this PR, or so that you can review the alternative and be sure it achieves your goal.

Also, since you brought the typing issues to my attention, how can I be sure to properly credit you? Your name in the commit? Or would you like me to apply the alternative to this PR so your name is paired in the commit itself?

**Comment by louiss0** [comment]
> Hi @louiss0, I have tagged you in the other PR, which I would like you to review.
> 
> > Are there any problems with the code I have written so far?
> 
> Well, yes. For example, your PR is documented as “Change Library to monorepo with better types” but it actually changes the library to not be a monorepo. The focus on typing is very unclear, and you made many changes totally unrelated to this. In your description, you talk about respecting the code, but then you add “_But I decided to change everything._” That makes it hard to see the gist of what you’re trying to accomplish.
> 
> In the PR that I tagged you in, it credits and attempts to apply the changes you suggested, without changing _everything_.
> 
> I very much want to partner with you here. It’s one of my favorite parts of open source. Again, it means a lot to me that you looked into the code and wanted to contribute. I wish I knew how to express that better.
> 
> I wanted to see if this PR could eventually take a shape that I could accept. It is still not there, which is why the alternative PR is up. The alternative tries to give you what you wanted without changing everything. There are some general issues stopping me from merging this PR as is. Most of them are side-effects from ‘changing everything’, and I will try to make them more explicit.
> 
> One nit: This PR makes it so users can’t contribute from a standard Node/NPM setup. I really like the accessibility of keeping things simple. That’s also one of the reasons I sometimes prefer JavaScript files over TypeScript files.
> 
> This may not seem relevant anymore, but at the time my alternative PR was made, your iterators were still not accepting generators. As my demos use iterators, I saw that you had changed them, but I had not seen a note from you explaining why you made this change. When I asked about it, you wrote that I can’t really use generators anymore, and I respectfully disagree with that. If this PR works with generators again, that deserves me giving it another look.
> 
> While I’ve listed out some issues with your PR, it’s only to make sure you know exactly why and how I’m hoping to get us the improvements you’ve brought up. As I wrote in my original reply to #1, I’m grateful you took the time to identify issues, and I want to get the typing exactly how you think it should be, as well as give you some credit. I’m also very sorry for any delays in my communication. I hope this feedback is helpful and not discouraging; either so that you can address the issues I’ve mentioned in this PR, or so that you can review the alternative and be sure it achieves your goal.
> 
> Also, since you brought the typing issues to my attention, how can I be sure to properly credit you? Your name in the commit? Or would you like me to apply the alternative to this PR so your name is paired in the commit itself?

I look at how the code works from the third pull request.
